### PR TITLE
`?url` and `?raw` suffix support

### PR DESCRIPTION
**Pull request by Thy3634**
Using `?url` suffix  to explicit URL Imports,
as well as `?raw` suffix to import .svg file as string.
#20  by gawlk

### CODE REVIEW

**Comment by jpkleemans** [comment]
Thanks @Thy3634
I'll look into it next week

**Comment by jpkleemans** [comment]
@Thy3634 Could you run the linter over your code? `npm run lint`

**Comment by Thy3634** [comment]
Failed. So I add a submit, and the code could be linted over now.

**Comment by jpkleemans** [comment]
> Failed. So I add a submit, and the code could be linted over now.

There are still some lint errors:
```
index.js:5:36: Missing space before function parentheses.
index.js:12:14: Missing space before function parentheses.
index.js:15:15: Missing space before function parentheses.
index.js:19:38: Strings must use singlequote.
index.js:20:21: Strings must use singlequote.
index.js:22:28: Strings must use singlequote.
index.js:26:20: Missing space before function parentheses.
index.js:30:38: Strings must use singlequote.
index.js:31:21: Strings must use singlequote.
index.js:39:36: Unexpected trailing comma.
index.js:40:11: Extra semicolon.
index.js:41:73: Extra semicolon.
```

**Comment by Thy3634** [comment]
Maybe git ignored something.
I add a commit via github directly. It should work.

File: index.js [hunk 1]
```diff
@@ -1,33 +1,47 @@
...
+        const svg = await fs.readFile(path, 'utf-8')
+        const optimizedSvg = svgo === false ? svg : optimizeSvg(svg, svgoConfig).data
+
+        let { code, ast, map } = compileTemplate({
+          id: JSON.stringify(id),
+          source: optimizedSvg,
+          filename: path,
+          transformAssetUrls: false
+        })
+        code = code.replace('export function render', 'function render')
```
Comment (jpkleemans) [inline index.js:41]: Why is this line needed?

File: index.js [hunk 2]
```diff
@@ -1,33 +1,47 @@
...
+        return { code: `${code}\nexport default render`, ast, map }
+      }
+      return `export default ${JSON.stringify(src)}`
```
Comment (jpkleemans) [inline index.js:44]: Please add an empty line before this return statement

File: index.js [hunk 3]
```diff
@@ -1,33 +1,47 @@
...
+      if (query === 'url') {
+        return path
+      } else if (query === 'raw') {
+        return await fs.readFile(path, 'utf-8')
+      }
+    },
+    async transform (src, id) {
+      if (!id.match(svgRegex)) {
+        return
+      }
```
Comment (jpkleemans) [inline index.js:29]: Please add an empty line after this if statement

File: index.js [hunk 4]
```diff
@@ -1,33 +1,47 @@
...
+      return id.match(svgRegex) ? id : null
+    },
     async load (id) {
-      const [path, parameter] = id.split('?')
-
-      if (!extname(path).startsWith('.svg') || parameter === 'url') {
-        return null
+      if (!id.match(svgRegex)) {
+        return
       }
```
Comment (jpkleemans) [inline index.js:18]: Please add an empty line after this if statement

File: index.js [hunk 5]
```diff
@@ -1,33 +1,47 @@
...
 
 module.exports = function svgLoader (options = {}) {
   const { svgoConfig, svgo } = options
 
+  const svgRegex = /\.svg(\?(raw|url|component))?$/
   return {
     name: 'svg-loader',
     enforce: 'pre',
-
+    resolveid (id) {
```
Comment (jpkleemans) [inline index.js:12]: Why is `resolveid` needed?

File: index.js [hunk 6]
```diff
@@ -1,33 +1,47 @@
...
-
-      return `${code}\nexport default render`
+      const [path, query] = id.split('?', 2)
+      if (query === 'url') {
+        return path
+      } else if (query === 'raw') {
+        return await fs.readFile(path, 'utf-8')
+      }
+    },
+    async transform (src, id) {
```
Comment (jpkleemans) [inline index.js:26]: Could you explain why you moved this code from `load()` to `transform()`?

File: index.js [hunk 7]
```diff
@@ -1,33 +1,47 @@
...
+      if (!id.match(svgRegex)) {
+        return
+      }
+      const [path, query] = id.split('?', 2)
+      if (query === 'component') {
```
Comment (jpkleemans) [inline index.js:31]: When no query suffixes are given, `component` should be the default behavior. Could you apply that?

**Review by jpkleemans — CHANGES_REQUESTED** [review]
I've placed some comments. Could you review them?

File: index.js (hunk 5 above)
Comment (Thy3634) [reply index.js:12]: If it cannot be resolved, return `null` so that Rollup displays an error.

File: index.js (hunk 7 above)
Comment (Thy3634) [reply index.js:31]: OK. But it will raise a type error.
```typescript
// vite/client.d.ts
declare module '*.svg' {
  const src: string
  export default src
}
```

File: index.js (hunk 1 above)
Comment (Thy3634) [reply index.js:41]: It is not necessary. But I believe less export would be better.

File: index.js (hunk 6 above)
Comment (Thy3634) [reply index.js:26]: In watch mode, the result of the `transform` hook is cached when rebuilding and the hook is only triggered again for a module id if either the code of the module has changed or a file has changed. 
[https://rollupjs.org/guide/en/#transform](url)

File: index.js (hunk 1 above)
Comment (jpkleemans) [reply index.js:47]: Agreed, but maybe then write it like this?
```js
code = code.replace('export function render', 'export default function render')
return { code, ast, map }
```

**Comment by jpkleemans** [comment]
Thanks for your changes! I've added 1 comment.

File: index.js (hunk 1 above)
Comment (Thy3634) [reply index.js]: OK. In addition, I found that this line will raise an error during build. Because the type is error. 

`transform` need return `{map?: string | SourceMap, ast? : ESTree.Program}` 
`compileTemplate` provide `{ast?: RootNode, map?: RawSourceMap}`

I don‘t know how to resolve it, so I just remove the `ast` and `map` .

**Review by jpkleemans — APPROVED** [review]

**Comment by jpkleemans** [comment]
Thanks for your work 👍
